src/Zcu.zig

```diff
@@ -3022,28 +3105,215 @@ pub const ResolvedReference = struct {
 };
 
 /// Returns a mapping from an `AnalUnit` to where it is referenced.
-/// TODO: in future, this must be adapted to traverse from roots of analysis. That way, we can
-/// use the returned map to determine which units have become unreferenced in an incremental update.
-pub fn resolveReferences(zcu: *Zcu) !std.AutoHashMapUnmanaged(AnalUnit, ResolvedReference) {
+/// If the value is `null`, the `AnalUnit` is a root of analysis.
+/// If an `AnalUnit` is not in the returned map, it is unreferenced.
+pub fn resolveReferences(zcu: *Zcu) !std.AutoHashMapUnmanaged(AnalUnit, ?ResolvedReference) {
```

Comment:
just to double check... this function is *only* called when a compile error occurs, right?

Reply:
Yep -- if there is an error in `failed_analysis`, we'll call `resolveReferences` from `Compilation`, which helps us (a) figure out the reference trace, and (b) determine if the failed `AnalUnit` is actually referenced in the latest update.

Except... mmm, I didn't actually put the call behind a condition. I'll do that.